trace/opencensus/trace/ext/django/middleware.py

```diff
@@ -0,0 +1,142 @@
+# limitations under the License.
+
+"""Django middleware helper to capture and trace a request."""
+import logging
+import threading
+
+from opencensus.trace.ext import utils
+from opencensus.trace.ext.django.config import settings
+from opencensus.trace.propagation.google_cloud_format import from_header
+
+_thread_locals = threading.local()
+
+HTTP_METHOD = '/http/method'
+HTTP_URL = '/http/url'
+HTTP_STATUS_CODE = '/http/status_code'
+
+REQUEST_SPAN_NAME = 'django_request_span'
+TRACER_THREAD_LOCAL_KEY = '_opencensus_django_request_tracer'
+
+_DJANGO_TRACE_HEADER = 'HTTP_X_CLOUD_TRACE_CONTEXT'
+
+log = logging.getLogger(__name__)
+
+
+def _get_django_request():
+    """Get Django request from thread local.
+
+    :rtype: str
+    :returns: Django request.
+    """
+    return getattr(_thread_locals, 'request', None)
+
+def _get_current_request_tracer():
+    """Get the current request tracer."""
+    return getattr(_thread_locals, TRACER_THREAD_LOCAL_KEY, None)
+
+def _set_django_labels(span, request):
+    """Set the django related labels."""
+    django_user = getattr(request, 'user', None)
+
+    if django_user is None:
+        return
+
+    user_id = django_user.getattr(django_user, 'pk', None)
+
+    if user_id is not None:
+        span.add_label('/django/user/id', user_id)
+
+    user_name = django_user.getattr(django_user, 'username', None)
+
+    if user_name is not None:
+        span.add_label('/django/user/name', user_name)
+
+def get_django_header():
+    """Get trace context header from django request headers.
+
+    :rtype: str
+    :returns: Trace context header in HTTP request headers.
+    """
+    request = _get_django_request()
+
+    header = request.META.get(_DJANGO_TRACE_HEADER)
+
+    return header
+
+class OpencensusMiddleware(object):
+    """Saves the request in thread local"""
+
+    def process_request(self, request):
+        """Called on each request, before Django decides which view to execute.
+
+        :type request: :class:`~django.http.request.HttpRequest`
+        :param request: Django http request.
+        """
+        # Add the request to thread local
+        _thread_locals.request = request
+
+        # Start tracing this request
+        tracer = settings.TRACER
+        sampler = settings.SAMPLER
+        reporter = settings.REPORTER
+
+        header = get_django_header()
+        span_context = from_header(header)
+        tracer = tracer(
+            span_context=span_context,
+            sampler=sampler(),
+            reporter=reporter())
+        tracer.start_trace()
+
+        try:
+            # Span name is being set at process_view
+            span = tracer.start_span()
+            span.add_label(label_key=HTTP_METHOD, label_value=request.method)
+            span.add_label(label_key=HTTP_URL, label_value=request.path)
+
+            # Add current span to thread local
+            setattr(_thread_locals, TRACER_THREAD_LOCAL_KEY, tracer)
+        except Exception:
```

Comment:
I prefer a broad except here to prevent unexpected errors which might fail the user's application.